### DISCUSSION TITLE
Control the display on your M5Stack Core2 with Toit

### DISCUSSION THREAD

**Discussion by kasperl** [toitlang/toit#5] [Show and tell]
# Control the display on your M5Stack Core2

The [M5Stack Core2](https://shop.m5stack.com/collections/stack-series/products/m5stack-core2-esp32-iot-development-kit) is a cool little IoT device based on the ESP32. It comes with a screen and it is a great device for tinkering with Toit.

# Steps

## 1. Get the source code

Clone the repositories in a suitable directory:

``` sh
git clone https://github.com/toitware/toit-m5stack-core2.git
git clone https://github.com/toitlang/toit.git
```

You should now have two additional directories: `toit-m5stack-core2/` and `toit/`.

## 2. Set up the build environment for Toit

Go into the `toit/` directory and follow the [instructions](https://github.com/toitlang/toit/#dependencies). You should be able to build the tools and the ESP32 firmware for a simple Toit:

``` sh
make tools
make esp32
```

If you run into other build issues ("error: loading 'build.ninja'"), you may not have set up your development environment correctly before you tried to build. Clean your build using `make clean` and make sure your `IDF_PATH` environment variable points to the right directory.

## 3. Install the package dependencies

The following command installs the package dependencies of the examples programs. Inside the `toit/` directory:

``` sh
build/host/sdk/bin/toit.pkg install --project-root=../toit-m5stack-core2/examples
```

## 4. Run the demo

Note: The core2 uses the QinHeng USB-to-serial chip, which needs a [driver-update on macos](https://github.com/WCHSoftGroup/ch34xser_macos).

The last step is to build the ESP32 firmware for the `tft_demo.toit` and flash it into your device. This is all done from the `toit/` directory:

``` sh
make flash ESP32_ENTRY=../toit-m5stack-core2/examples/tft_demo.toit ESP32_PORT=/dev/ttyACM0
```

It is often useful to see the serial output from the device while it is running and you can use the `toit` command-line interface for that too:

``` sh
toit serial monitor
```

or you can stick to simpler alternatives like `cat`:

``` sh
stty -F /dev/ttyACM0 115200
cat /dev/ttyACM0
```

## 5. Play!

Edit the `toit-m5stack-core2/examples/tft_demo.toit` or make a copy of it and try to change some of the text on the screen. Maybe something close to [line 42](https://github.com/toitware/toit-m5stack-core2/blob/main/examples/tft_demo.toit#L42) would be a good start!

For more advanced changes to the example you probably want to take a look at [how the display code works](https://docs.toit.io/language/sdk/display) or look at the examples in [the toit-color-tft repo](https://github.com/toitware/toit-color-tft).

Once you have changed the example, you need to build and run the demo again from step 4. Have fun!

# Picture or it didn't happen

![PXL_20211118_095534785](https://user-images.githubusercontent.com/2156198/142585893-621cbe38-648f-4465-af26-225a00ad6dc5.jpg)

**Comment by imliubo** [comment]
Hi,
It seems that I cannot compile for esp32 correctly. When I execute **make esp32**, cmake will report the following error. 
```shell
Compiling the CXX compiler identification source file "CMakeCXXCompilerId.cpp" failed.
Compiler: /usr/bin/c++ 
Build flags: ;-m32
Id flags:  

The output was:
1
/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/9/libstdc++.so when searching for -lstdc++
/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/9/libstdc++.a when searching for -lstdc++
/usr/bin/ld: cannot find -lstdc++
collect2: error: ld returned 1 exit status


Determining if the CXX compiler works failed with the following output:
Change Dir: /path/to/toit/build/ia32/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/ninja cmTC_e43f4 && [1/2] Building CXX object CMakeFiles/cmTC_e43f4.dir/testCXXCompiler.cxx.o
[2/2] Linking CXX executable cmTC_e43f4
FAILED: cmTC_e43f4 
: && /usr/bin/c++  -m32   CMakeFiles/cmTC_e43f4.dir/testCXXCompiler.cxx.o  -o cmTC_e43f4   && :
/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/9/libstdc++.so when searching for -lstdc++
/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/9/libstdc++.a when searching for -lstdc++
/usr/bin/ld: cannot find -lstdc++
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```
I checked the cmake file(toolchains/ia32.cmake) carefully. I think it is caused by the "-m32" parameter.So I deleted the "-m32" parameter, it can compile success:
```shell
make esp32 ESP32_ENTRY=../toit-m5stack-core2/examples/tft_demo.toit
...
App "toit" version: fc7362d-dirty
LD build/esp32/toit.elf
esptool.py v3.1-dev
Merged 2 ELF sections
To flash all build output, run 'make flash' or:
python /path/to/esp-idf-toit/components/esptool_py/esptool/esptool.py --chip esp32 --port /dev/ttyUSB0 --baud 921600 --before default_reset --after hard_reset write_flash -z --flash_mode dio --flash_freq 40m --flash_size detect 0xd000 
/path/to/toit/build/esp32/ota_data_initial.bin 0x1000 
/path/to/toit/build/esp32/bootloader/bootloader.bin 0x10000 
/path/to/toit/build/esp32/toit.bin 0x8000 
/path/to/toit/build/esp32/partitions.bin
```
But when I try to burn to esp32, it doesn't seem to run correctly.
```shell
Rebooting...
[flash reg] address 0x3f43c000, size 0x00134000
Guru Meditation Error: Core  0 panic'ed (LoadProhibited). Exception was unhandled.

Core  0 register dump:
PC      : 0x4018aa06  PS      : 0x00060130  A0      : 0x801054b8  A1      : 0x3ffbbb60  
A2      : 0x3ffd000d  A3      : 0x3f401bb4  A4      : 0x00000000  A5      : 0x00060523  
A6      : 0x3ffc8000  A7      : 0x00008000  A8      : 0x00000000  A9      : 0x00000000  
A10     : 0x00000000  A11     : 0x00000000  A12     : 0x3ffcf9f0  A13     : 0x00000120  
A14     : 0x000004e8  A15     : 0x3ffcfb0c  SAR     : 0x00000009  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x00000000  LBEG    : 0x4000c2e0  LEND    : 0x4000c2f6  LCOUNT  : 0x00000000  

Backtrace:0x4018aa03:0x3ffbbb60 0x401054b5:0x3ffbbb80 0x401054d3:0x3ffbbba0 0x401057cd:0x3ffbbbc0 0x40111441:0x3ffbbbe0 0x401114c2:0x3ffbbc00 0x401128a5:0x3ffbbc20 0x4011317d:0x3ffbbc60 0x40113277:0x3ffbbf20 0x400f9c27:0x3ffbbf40 0x4019516d:0x3ffbbf60


ELF file SHA256: bed0f0023456543a
```
Do you have any advice? Am I missing something?

**Reply by kasperl** [reply]
Hi! 

It is important that you use a 32-bit build of the compiler to produce the correct executable image for the ESP32. It is because the memory layout used in the compiler needs to be compatible with the layout on the ESP32 (pointer sizes, etc.) so the ESP32 doesn't have to shuffle bits around at start up. I can imagine us fixing this by simulating a 32-bit memory model on a 64-bit machine that runs the compiler, but it is a bit of work.

Instead of removing the `-m32` option, I think we need to find a way to add 32-bit executable support to your machine. Which operating system / distribution are you running?

Cheers,
Kasper

**Reply by imliubo** [reply]
Thanks for reply!
My OS info:
```shell
No LSB modules are available.
Distributor ID:	Linuxmint
Description:	Linux Mint 20.2
Release:	20.2
Codename:	uma
```

**Reply by kasperl** [reply]
I think the right packages to install are typically named `gcc-multilib` and `g++-multilib`. If you use `apt-get`, it is probably something like this:

``` sh
sudo apt-get install gcc-multilib g++-multilib
```

If this solves your issues, I will go ahead and update the setup instructions to cover this.

**Reply by imliubo** [reply]
Yes, it works!
```shell
clearing RTC memory: RTC memory is in inconsistent state
[flash reg] address 0x3f42c000, size 0x00134000
Hello, World!
E (442) Toit: VM exited, going into deep sleep.
```
Yep, we need add this to instruction.

**Reply by kasperl** [reply]
I've updated this tutorial with the extra information. I will also update the general instructions.

Thanks!

**Comment by kasperl** [comment]
The general instructions have been updated. We will continue to improve them. Thanks again!

Did you manage to get something shown on your M5Stack Core2 display from Toit code, @imliubo?

**Reply by imliubo** [reply]
Yes, the LCD works OK, I am reading toit's documentation to get familiar with the syntax of toit.

**Comment by imliubo** [comment]
Hi,
After spending a little time, MPU6886 seems to be able to work. Once you are familiar with toit, development is still very fast. I think this is the **future** of embedded development. Your work is excellent!
I am learning how to register mpu6886 package, hope to try it tomorrow, it is a bit late today.

Thanks for your help!

https://user-images.githubusercontent.com/26544331/142768842-b5c1e063-fbfa-4b54-9760-d03d1d3e38f5.mp4

**Reply by floitsch** [reply]
Nice!

Let me know if you need help with the package.
If you want to I'm also happy to do a review before (or after) publishing the package.

**Reply by imliubo** [reply]
Thanks!
I upload a very sample version to this repo: https://github.com/imliubo/mpu6886-toit

**Reply by kasperl** [reply]
I'll let @floitsch help with the packaging, but you may want to consider using https://pkg.toit.io/package/github.com%2Fandersjohnsen%2Fahrs for the AHRS implementation?

**Reply by imliubo** [reply]
Wow, I did't notice that, I will use this package.Has any example?

**Reply by kasperl** [reply]
No, it looks like the package could use an example. I found one use of it here:

https://github.com/andersjohnsen/stabilizer/blob/main/stabilizer.toit#L52

**Reply by floitsch** [reply]
Very nice package. Only had minor nits (and I'm trying to be really nitpicky).
https://github.com/imliubo/mpu6886-toit/pull/1
Thanks!

**Reply by floitsch** [reply]
@imliubo The last remaining steps for the package: add a tag "v1.0.0", and then publish it here: https://pkg.toit.io/publish
Looking forward to seeing it available for everyone :)

**Reply by imliubo** [reply]
Done.I can see [here](https://github.com/toitware/registry/blob/main/packages/github.com/imliubo/mpu6886-toit/1.0.0/desc.yaml)

**Reply by kasperl** [reply]
Awesome: https://pkg.toit.io/package/github.com%2Fimliubo%2Fmpu6886-toit.

**Reply by kasperl** [reply]
Hi @imliubo! I'd love to get your feedback on the live reloading support we added in https://github.com/toitlang/jaguar. It might be a much faster way of iterating on drivers like the ones you already wrote.

**Comment by imliubo** [comment]
Touch test.

[package link](https://github.com/imliubo/ft63xx-toit)

https://user-images.githubusercontent.com/26544331/143904825-24d4f187-3d5f-47c6-af47-7f012aeedb58.mp4

**Reply by floitsch** [reply]
Really nice.

**Reply by un01s** [reply]
Finally run this package on my M5Core2. Thanks.

[toit on m5core2](https://un01s.github.io/posts/20220115/)

**Reply by floitsch** [reply]
Thanks for the write-up.
I wrote a PR to fix the yaml file: https://github.com/imliubo/ft63xx-toit/pull/1

**Comment by lutzbickhardt** [comment]
Very nice, 

runs on macOS and Ubuntu; had to change: 
a) build/toitpkg pkg install --project-root=../toit-m5stack-core2/examples to
A) toit pkg install --project-root=../toit-m5stack-core2/examples

b) in the Makefile 
B1) python to python3 (Mac) and 
B2) go install ... @v.0.0.0... to @latest (MacOS) resp. omit at all (Ubuntu)

All depending potentially/probably on the versions used (Mac Catalina 10.15.7, Ubuntu 20.04) 
also had to upgrade to go 1.17 (mac).

Also IDF_PATH has always to be set (not only when using non-standard path).

Also runs very good with jaguar.

Question: it does not work (as expected) with M5Stack Basic, Fire, ... (which are not core2) and M5Sticks. They are all based on ESP32, but need a different package. Is that planned?

Thanks a lot!

**Reply by floitsch** [reply]
> A) toit pkg install --project-root=../toit-m5stack-core2/examples

This indicates that you are using the toit.io version. The open-source code doesn't contain any `toit` executable (yet).
Especially for the package manager there is no difference, so it shouldn't matter.

> B1) python to python3 (Mac) and

Python...
In theory we don't use Python, but mbedtls (one of our dependencies) does. I'm not sure if Python is also needed when creating the image. If not, we might be able to drop the Python dependency in the future.

> B2) go install ... @v.0.0.0... to @latest (MacOS) resp. omit at all (Ubuntu)

That's curious. I'm not a Go expert, but I thought that Go 1.16+ deals with that correctly. I might need to have another look at this... :(

>  ... but need a different package. Is that planned?

The M5Stack Basic (and probably) others work fine with Toit, but don't have specific packages yet.
I thought of this myself, and planned on discussing this with others: the idea would be to have packages for the most common dev-boards. They could then be referenced on https://github.com/toitlang/toit/wiki/Devices.
I guess it would be nice to have some conventions for these packages (naming, description, ...) but that should be easy to do.

**Reply by lutzbickhardt** [reply]
Hi Florian,

you were right I used the toit.io version on Ubuntu and macOS (did not really remember I had installed it!). And I was not aware that there is a toit.io and a toitlang version.

However,
-build/toitpkg pkg install --project-root=../toit-m5stack-core2/examples 
does not work because it should read
-build/host/sdk/bin/toitpkg pkg install --project-root=../toit-m5stack-core2/examples.

Repeated the instructions on a newly setup Ubuntu 21.10; worked as described but needed to : 
- sudo apt install g++-multilib gcc-multilib python3-pip golang
- sudo apt remove cmake (version 3.18, but need >= 3.19)
- sudo snap install cmake --classic (version 3.22).

Then I could continue to install m5stack-core2 package, download jaguar and use VS Code without installing the extension for toit.io.
So far, so good.

Regarding the m5stack basic etc. I will write a separate comment; so far the examples only work with core2 and core2-AWS modules.

**Reply by floitsch** [reply]
Thanks. We moved the executable some time ago and forgot to update the post.
It's fixed now.
